File: tests/typesafety/python/carrier_protocol.py
# expect: passes
# run: yes
# title: A Protocol accepts any carrier with the right method, no inheritance asked
# improves: carrier_inheritance
from typing import Protocol


class Carrier(Protocol):
    def quote_pence(self, kilograms: float) -> int: ...


class RoyalPost:
    def quote_pence(self, kilograms: float) -> int:
        return int(kilograms * 120)


class VillageCourier:
    def quote_pence(self, kilograms: float) -> int:
        return 90


def cheapest(carriers: list[Carrier], kilograms: float) -> int:
    return min(carrier.quote_pence(kilograms) for carrier in carriers)


assert cheapest([RoyalPost(), VillageCourier()], 9.5) == 90

File: tests/typesafety/typescript/carrier_protocol.ts
// expect: passes

type Carrier = {
  quotePence(kilograms: number): number;
};

class RoyalPost {
  quotePence(kilograms: number): number {
    return Math.floor(kilograms * 120);
  }
}

const villageCourier: Carrier = {
  quotePence: () => 90,
};

export function cheapest(carriers: readonly Carrier[], kilograms: number): number {
  return Math.min(...carriers.map((carrier) => carrier.quotePence(kilograms)));
}

export const quote = cheapest([new RoyalPost(), villageCourier], 9.5);
